### DISCUSSION TITLE
Tagged PDFs?

### DISCUSSION THREAD

**Discussion by plweil** [spatie/browsershot#799] [Q&A] [answered]
Is anyone able to generate tagged PDFs (essential for accessibility) with Browsershot/Puppeteer? I don't know if my brain is playing tricks on me, but I could swear that PDFs generated in the app I'm developing were found to be tagged by Acrobat, but now they're not. I saw an option for this in Puppeteer, but either I don't know how to apply it, or it doesn't work. I've also seen some discussions regarding Chromium support for tagged PDFs, but again, nothing that clearly indicated whether this actually works or how to get it to work. With all of these pieces of technology working together, which one is responsible for supporting this?

**Comment by ntaylor-86** [comment]
Does adding the below flag to chromium work?

```php
Browsershot::html('Foo')
  ->addChromiumArguments([
      'export-tagged-pdf',
  ]);
```

Sources:

https://spatie.be/docs/browsershot/v2/requirements#content-pass-custom-arguments-to-chromium

https://github.com/puppeteer/puppeteer/pull/5337

**Comment by plweil** [comment]
Thanks @ntaylor-86. No, it does not. I thought Chromium is supposed to[ generate tagged PDFs by default](https://blog.chromium.org/2020/07/using-chrome-to-generate-more.html) now. I don't know what's breaking this — Chromium, Puppeteer, Browsershot.... — but it's a real shame considering how essential tags are for PDF accessibility. And I'm surprised that no one has brought this up here previously. I can't be the only one concerned about generating tagged PDFs.

**Reply by ntaylor-86** [reply]
It would be Chromium that would be breaking this. Check what version of Chromium, Puppeteer is using under the hood, then see what flags you have to use for that particular version.

**Comment by ntaylor-86** [comment]
I just tested my local version of chrome using puppeteer (HeadlessChrome/105.0.5173.0);

![image](https://github.com/spatie/browsershot/assets/28693782/7dff356d-4f62-4007-a431-161a2eaf0fe8)

Which lines up with the version of Puppeteer I've got installed `"puppeteer": "^15.5.0"` (https://pptr.dev/chromium-support)

Using the Chromium Blog as an example, I got Browsershot to save the pdf.

```php
Browsershot::url('https://blog.chromium.org/2020/07/using-chrome-to-generate-more.html')->savePdf('ChromiumBlog.pdf');
```

I didn't know the best way to check if the pdf was tagged, I found Read PDF Metadata (https://www.pdfyeah.com/view-pdf-metadata/)
Which returned the below results when I uploaded my Browsershot example.

```
Creator:        Chromium
Producer:       Skia/PDF m105
CreationDate:   Tue Dec 19 23:49:20 2023 UTC
ModDate:        Tue Dec 19 23:49:20 2023 UTC
Tagged:         yes
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          2
Encrypted:      no
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      136086 bytes
Optimized:      no
PDF version:    1.4
```

Looks like this version of Chromium is generating tags without any additional flags.

**Reply by plweil** [reply]
I think the best way (in addition to the metadata check) to check is to open the PDF in Acrobat and run an accessibility check. I have Puppeteer 21.6.1. I can't seem to run `const page = await browser.newPage();`. I get
```
Uncaught TypeError: browser.newPage is not a function
    at REPL18:1:53
```

At any rate, checking the metadata of my PDF indicates that it is missing tags. I'm a little stumped as to why yours does and mine doesn't.  I am generating the PDF like this in a Laravel controller:

```php
        $html = view('auth.reports.test-new-pdf', $data)->render();

        $pdfContent = Browsershot::html($html)
            ->setNodeBinary(config('constants.node_bin').'node')
            ->setNpmBinary(config('constants.node_bin').'npm')
            ->showBackground()
            ->pdf();
            
          return response($pdfContent)->withHeaders([
            'content-type' => 'application/pdf',
        ]);          
```

[EDIT] I have also tried savePdf(), but still no tags.
Looks as though Puppeteer installed [Chrome for Testing](https://developer.chrome.com/blog/chrome-for-testing/) 119.0.6045.105

**Reply by ntaylor-86** [reply]
I just created a new Laravel project. Installed Puppeteer 21.6.1 and Browsershot 3.6

Puppeteer is now using `HeadlessChrome/119.0.6045.105`

Tried my Browsershot test of the Chromium Blog again.

```
Creator:        Chromium
Producer:       Skia/PDF m119
CreationDate:   Wed Dec 20 01:15:09 2023 UTC
ModDate:        Wed Dec 20 01:15:09 2023 UTC
Tagged:         no
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          3
Encrypted:      no
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      120672 bytes
Optimized:      no
PDF version:    1.4
```

Now there are no tags in the PDF. It's gotta be the Chromium version causing this.

**Reply by plweil** [reply]
That could be good news, sort of. Maybe I should try reinstalling an older version of Puppeteer (and with it, Chrome). Come to think of it, I could swear that I was getting tagged pdfs a couple of months ago. Maybe it's something about when they moved to that "Chrome for Testing".

**Reply by ntaylor-86** [reply, marked as answer]
Give this a try.

Edit this file:

`vendor/spatie/browsershot/src/Browsershot.php`

Modify the `createPdfCommand()` method:

and add the line `$command['options']['tagged'] = true;`

e.g.

```php
  public function createPdfCommand($targetPath = null): array
  {
      $url = $this->getFinalContentsUrl();

      $options = [];

      if ($targetPath) {
          $options['path'] = $targetPath;
      }

      $command = $this->createCommand($url, 'pdf', $options);

      /**
       * @plweil 
       * 
       * Added this line below
       * 
       */
      $command['options']['tagged'] = true;

      if ($this->showBackground) {
          $command['options']['printBackground'] = true;
      }

      if ($this->transparentBackground) {
          $command['options']['omitBackground'] = true;
      }

      if ($this->scale) {
          $command['options']['scale'] = $this->scale;
      }

      return $command;
  }
```

After adding the tagged option, this is now the PDF metadata:

```
Creator:        Chromium
Producer:       Skia/PDF m119
CreationDate:   Wed Dec 20 01:57:51 2023 UTC
ModDate:        Wed Dec 20 01:57:51 2023 UTC
Tagged:         yes
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          3
Encrypted:      no
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      137373 bytes
Optimized:      no
PDF version:    1.4
```

**Reply by plweil** [reply]
Yep, that works! Shouldn't one be able to add this option when you run the `pdf()` command? Something like this?:
```
        $pdfContent = Browsershot::html($html)
            ->setNodeBinary(config('constants.node_bin').'node')
            ->setNpmBinary(config('constants.node_bin').'npm')
            ->showBackground()
            ->setOption('tagged', true)
            ->pdf();
```

**Reply by ntaylor-86** [reply]
I've created a pull request #804 that adds a method to achieve this.

**Reply by plweil** [reply]
Thanks. I suppose there's something about the newer version of Chromium that either turns this off by default or somehow negates this. I'm pretty sure that it's supposed to be on by default.
